siad/daemon.go

```diff
@@ -107,13 +108,36 @@ func createSubdirs(rootDir string) error {
 }
 
 // TODO: move this to the state module?
-func (d *daemon) relayBlock(conn modules.NetConn) error {
+func (d *daemon) acceptBlock(conn modules.NetConn) error {
 	var b consensus.Block
 	err := conn.ReadObject(&b, consensus.BlockSizeLimit)
 	if err != nil {
 		return err
 	}
-	return d.gateway.RelayBlock(b)
+
+	err = d.state.AcceptBlock(b)
+	if err == consensus.ErrOrphan {
+		go d.gateway.Synchronize(conn.Addr())
+		return err
+	} else if err != nil {
+		return err
+	}
+
+	// Check if b is in the current path.
+	height, exists := d.state.HeightOfBlock(b.ID())
+	if !exists {
+		if consensus.DEBUG {
+			panic("could not get the height of a block that did not return an error when being accepted into the state")
+		}
+		return errors.New("state malfunction")
+	}
+	currentPathBlock, exists := d.state.BlockAtHeight(height)
+	if !exists || b.ID() != currentPathBlock.ID() {
+		return errors.New("block added, but it does not extend the state height")
```

Comment:
This looks like a race condition, because multiple threads could call AcceptBlock by the time you get to here.

Maybe we should have the state return ErrNotHeaviestChain or something. We use locks a bunch and I think it's causing a lot of problems; poor design choice I guess.